SCI: (KQ6WinCD) - fix hires drawing glitches

athrxx opened the pull request:
This should fix issues 1- 3 from bug ticket no. 15594.

I have added redrawing functionality for the hires gfx from the original interpreter. I guess the original only needed that code to handle WM_PAINT messages (= full window redraw). It also does actually call a redraw in kDisposeWindow, the same way I implemented it here, to refresh the inventory. But that is only needed for the speech+text mode which isn't supported in the original. Maybe they ditched the mode only after they had already implemened that, who knows...

I have also improved the workaround for the out-of-bounds portraits to catch the case from the bug ticket.

sluicebox commented on engines/sci/graphics/paint16.cpp:
```diff
@@ -131,31 +153,41 @@ void GfxPaint16::drawCel(GfxView *view, int16 loopNo, int16 celNo, const Common:
+	// window background when recieving WM_PAINT messages (which is irrelevant for us, since that happens in the backend) and the other is
+	// redrawing the inventory after displaying a text window over it. This only happens in mixed speech+text mode which does not even exist
+	// in the original. We do have that mode as a ScummVM feature, though. That's why we have that code, to be able to refresh the inventory.
+	// We also check if the portrait is drawn outside the viewport boundaries (happens in the inofficial mixed speech+text mode) and set
```
"unofficial"

bluegr replied:
```suggestion
	// We also check if the portrait is drawn outside the viewport boundaries (happens in the unofficial mixed speech+text mode) and set
```

sluicebox commented on engines/sci/graphics/paint16.cpp:
```diff
@@ -131,31 +153,41 @@ void GfxPaint16::drawCel(GfxView *view, int16 loopNo, int16 celNo, const Common:
-// This is used as replacement for drawCelAndShow() when hires-cels are drawn to
-// screen. Hires-cels are available only SCI 1.1+.
-void GfxPaint16::drawHiresCelAndShow(GuiResourceId viewId, int16 loopNo, int16 celNo, uint16 leftPos, uint16 topPos, byte priority, uint16 paletteNo, reg_t upscaledHiresHandle, uint16 scaleX, uint16 scaleY) {
+// This is used as replacement for drawCelAndShow() when hires-cels are drawn to screen. Hires-cels are available only SCI 1.1+.
```
"SCI 1.1+" => "SCI 1.1", since this is SCI16-only code. (I know it's an existing comment)

sluicebox commented:
LGTM; I haven't tested this but it's well commented and, as always, I'm convinced you know what you're doing =)

Thanks for reverse engineering these edge cases! I'll let @bluegr take a look and merge it, so that I can blame him if it crashes later.

bluegr commented on engines/sci/graphics/paint16.cpp:
```diff
@@ -131,31 +153,41 @@ void GfxPaint16::drawCel(GfxView *view, int16 loopNo, int16 celNo, const Common:
 	view->draw(celRect, clipRect, clipRect, loopNo, celNo, priority, paletteNo, true);
+
+	// The original KQ6WinCD interpreter saves the hires drawing information in a linked list. There are two use cases, one is redrawing the
+	// window background when recieving WM_PAINT messages (which is irrelevant for us, since that happens in the backend) and the other is
```
```suggestion
	// window background when receiving WM_PAINT messages (which is irrelevant for us, since that happens in the backend) and the other is
```

bluegr commented on engines/sci/graphics/paint16.cpp:
```diff
@@ -131,31 +153,41 @@ void GfxPaint16::drawCel(GfxView *view, int16 loopNo, int16 celNo, const Common:
+
+	// The original KQ6WinCD interpreter saves the hires drawing information in a linked list. There are two use cases, one is redrawing the
+	// window background when recieving WM_PAINT messages (which is irrelevant for us, since that happens in the backend) and the other is
+	// redrawing the inventory after displaying a text window over it. This only happens in mixed speech+text mode which does not even exist
```
```suggestion
	// redrawing the inventory after displaying a text window over it. This only happens in mixed speech+text mode, which does not even exist
```

bluegr commented on engines/sci/graphics/paint16.cpp:
```diff
@@ -131,31 +153,41 @@ void GfxPaint16::drawCel(GfxView *view, int16 loopNo, int16 celNo, const Common:
-// This is used as replacement for drawCelAndShow() when hires-cels are drawn to
-// screen. Hires-cels are available only SCI 1.1+.
-void GfxPaint16::drawHiresCelAndShow(GuiResourceId viewId, int16 loopNo, int16 celNo, uint16 leftPos, uint16 topPos, byte priority, uint16 paletteNo, reg_t upscaledHiresHandle, uint16 scaleX, uint16 scaleY) {
+// This is used as replacement for drawCelAndShow() when hires-cels are drawn to screen. Hires-cels are available only SCI 1.1+.
```
```suggestion
// This is used as replacement for drawCelAndShow() when hires cels are drawn to screen. Hires cels are available only in SCI 1.1.
```

athrxx replied:
I didn't add that comment. Why would I say `SCI 1.1` when I am just talking about KQ6Win. But why not change it if we're at it anyway...

bluegr commented on engines/sci/graphics/paint16.cpp:
```diff
@@ -131,31 +153,41 @@ void GfxPaint16::drawCel(GfxView *view, int16 loopNo, int16 celNo, const Common:
+	byte *memoryPtr = _segMan->getHunkPointer(hiresHandle);
+	if (!memoryPtr) {
+		// The original KQ6WinCD interpreter does not treat this as an error, it just skips the hires drawing. It happens all the time
+		// when attempting to redraw the hires cels from the _hiresDrawObjs chain, we're supposed to just skip the invalidated items.
```
```suggestion
		// when attempting to redraw the hires cels from the _hiresDrawObjs chain. We're supposed to just skip the invalidated items.
```

bluegr commented on engines/sci/graphics/paint16.cpp:
```diff
@@ -619,4 +654,34 @@ void GfxPaint16::kernelPortraitShow(const Common::String &resourceName, Common::
+			continue;
+
+		// WORKAROUND for vertically misplaced hires portraits in mixed speech+text mode in KQ6CD. If we have
+		// an entry which is flagged as needing a workaround we set the notification for bitsShow() here.
```
```suggestion
		// an entry which is flagged as needing a workaround, we set the notification for bitsShow() here.
```

bluegr commented on engines/sci/graphics/paint16.h:
```diff
@@ -96,6 +98,16 @@ class GfxPaint16 {
 	// true means make EGA picture drawing visible
 	bool _EGAdrawingVisualize;
+
+	// The original KQ6WinCD interpreter saves the hires drawing information in a linked list. There are two use cases, one is redrawing the
```
```suggestion
	// The original KQ6WinCD interpreter saves the hires drawing information in a linked list. There are two use cases: one is redrawing the
```

bluegr commented on engines/sci/graphics/paint16.h:
```diff
@@ -96,6 +98,16 @@ class GfxPaint16 {
 	bool _EGAdrawingVisualize;
+
+	// The original KQ6WinCD interpreter saves the hires drawing information in a linked list. There are two use cases, one is redrawing the
+	// window background when recieving WM_PAINT messages (which is irrelevant for us, since that happens in the backend) and the other is
```
```suggestion
	// window background when receiving WM_PAINT messages (which is irrelevant for us, since that happens in the backend) and the other is
```

bluegr commented on engines/sci/graphics/paint16.h:
```diff
@@ -96,6 +98,16 @@ class GfxPaint16 {
+
+	// The original KQ6WinCD interpreter saves the hires drawing information in a linked list. There are two use cases, one is redrawing the
+	// window background when recieving WM_PAINT messages (which is irrelevant for us, since that happens in the backend) and the other is
+	// redrawing the inventory after displaying a text window over it. This only happens in mixed speech+text mode which does not even exist
```
```suggestion
	// redrawing the inventory after displaying a text window over it. This only happens in mixed speech+text mode, which does not even exist
```

bluegr reviewed (approved):
LGTM! Nice work :)
Shouldn't we add a game ID check in all that hires SCI11 code? If no other games are affected, we can skip the check

athrxx commented:
> LGTM! Nice work :) Shouldn't we add a game ID check in all that hires SCI11 code? If no other games are affected, we can skip the check

Thanks! KQ6Win is the only game to reach the relevant code paths. No other game will call `drawHiresCelAndShow()`. And if that method is not called, the draw chain (`_hiresDrawObjs`) will remain a nullptr. Which means, the rest of the code will not be executed either. Unless I made a mistake somwhere...

sluicebox commented:
> > LGTM! Nice work :) Shouldn't we add a game ID check in all that hires SCI11 code? If no other games are affected, we can skip the check
> 
> Thanks! KQ6Win is the only game to reach the relevant code paths. No other game will call `drawHiresCelAndShow()`. And if that method is not called, the draw chain (`_hiresDrawObjs`) will remain a nullptr. Which means, the rest of the code will not be executed either. Unless I made a mistake somwhere...

When I was reviewing I was going to something similar to @bluegr but then I looked close and saw that the linked list implementation meant a nullptr wasn't going to crash it. So while I don't think it needs to change, that's two people who this raised a question for. Although for me it was on the kDisposeWindow change because that gets called by so many other games. @athrxx it's up to you, but maybe a game/version check there for more the sake of clarity and documentation? I don't feel strongly about this, but when two of us do the same double take, maybe it could be clearer..

athrxx commented:
I have added two sanity checks at locations which might not be so obvious (including kDisposeWindow ).

And I also renamed a var that I had renamed previously, because the name made no sense. But I managed to rename it to something that also made no sense :sweat_smile: .
